clients/filesystem-fuse/src/default_raw_filesystem.rs

```diff
@@ -0,0 +1,393 @@
+            Self::ROOT_DIR_FILE_ID,
+            Self::ROOT_DIR_NAME,
+        )
+        .await;
+        self.fs.init().await
+    }
+
+    async fn get_file_path(&self, file_id: u64) -> Result<String> {
+        let file_entry = self.get_file_entry(file_id).await;
+        Ok(file_entry?.path)
+    }
+
+    async fn valid_file_handle_id(&self, file_id: u64, fh: u64) -> Result<()> {
+        let fh_file_id = self
+            .opened_file_manager
+            .get(fh)
+            .ok_or(Errno::from(libc::EBADF))?
+            .lock()
+            .await
+            .file_stat
+            .file_id;
+
+        (file_id == fh_file_id)
+            .then_some(())
+            .ok_or(Errno::from(libc::EBADF))
+    }
+
+    async fn stat(&self, file_id: u64) -> Result<FileStat> {
+        let file_entry = self.get_file_entry(file_id).await?;
+        let mut file_stat = self.fs.stat(&file_entry.path).await?;
+        file_stat.set_file_id(file_entry.parent_file_id, file_entry.file_id);
+        Ok(file_stat)
+    }
+
+    async fn lookup(&self, parent_file_id: u64, name: &str) -> Result<FileStat> {
+        let parent_file_entry = self.get_file_entry(parent_file_id).await?;
+        let mut file_stat = self.fs.lookup(&parent_file_entry.path, name).await?;
+        // fill the file id to file stat
+        self.resolve_file_id_to_filestat(&mut file_stat, parent_file_id)
+            .await;
+        Ok(file_stat)
+    }
+
+    async fn read_dir(&self, file_id: u64) -> Result<Vec<FileStat>> {
+        let file_entry = self.get_file_entry(file_id).await?;
+        let mut child_filestats = self.fs.read_dir(&file_entry.path).await?;
+        for file in child_filestats.iter_mut() {
+            self.resolve_file_id_to_filestat(file, file.file_id).await;
+        }
+        Ok(child_filestats)
+    }
+
+    async fn open_file(&self, file_id: u64, flags: u32) -> Result<FileHandle> {
+        self.open_file_internal(file_id, flags, FileType::RegularFile)
+            .await
+    }
+
+    async fn open_dir(&self, file_id: u64, flags: u32) -> Result<FileHandle> {
+        self.open_file_internal(file_id, flags, FileType::Directory)
+            .await
+    }
+
+    async fn create_file(&self, parent_file_id: u64, name: &str, flags: u32) -> Result<FileHandle> {
+        let parent_file_entry = self.get_file_entry(parent_file_id).await?;
+        let mut opened_file = self
+            .fs
+            .create_file(&parent_file_entry.path, name, OpenFileFlags(flags))
+            .await?;
+
+        opened_file.set_file_id(parent_file_id, self.next_file_id());
+
+        // insert the new file to file entry manager
+        self.insert_file_entry_locked(
+            parent_file_id,
+            opened_file.file_stat.file_id,
+            &opened_file.file_stat.path,
+        )
+        .await;
+
+        // put the file to the opened file manager
+        let opened_file = self.opened_file_manager.put(opened_file);
```

Comment:
I'm confused by why not return the file handle in the first opened_file.  after reading the code, I know you fill the file handle id, it's hacky here

Reply:
add comments and rename parameters name